__tests__/Auth0Client.test.ts

```diff
@@ -173,28 +177,389 @@ describe('Auth0Client', () => {
     });
   });
 
+  describe('loginWithPopup', () => {
+    it('encodes state with random string', async () => {
+      const auth0 = setup();
+
+      await loginWithPopup(auth0);
+
+      const [[url]] = (<jest.Mock>mockWindow.open).mock.calls;
+      assertUrlEquals(url, 'auth0_domain', '/authorize', {
+        state: TEST_STATE,
+        nonce: TEST_NONCE
+      });
+    });
+
+    it('creates `code_challenge` by using `utils.sha256` with the result of `utils.createRandomString`', async () => {
+      const auth0 = setup();
+
+      await loginWithPopup(auth0);
+
+      const [[url]] = (<jest.Mock>mockWindow.open).mock.calls;
+      assertUrlEquals(url, 'auth0_domain', '/authorize', {
+        code_challenge: TEST_CODE_CHALLENGE,
+        code_challenge_method: 'S256'
+      });
+    });
+
+    it('should log the user in with a popup and redirect using a default redirect URI', async () => {
+      const auth0 = setup({ leeway: 10, redirect_uri: null });
+
+      await loginWithPopup(auth0, {
+        connection: 'test-connection',
+        audience: 'test'
+      });
+
+      expect(mockWindow.open).toHaveBeenCalled();
+      const [[url]] = (<jest.Mock>mockWindow.open).mock.calls;
+      assertUrlEquals(url, 'auth0_domain', '/authorize', {
+        redirect_uri: 'http://localhost',
+        client_id: TEST_CLIENT_ID,
+        scope: TEST_SCOPES,
+        response_type: 'code',
+        response_mode: 'web_message',
+        state: TEST_STATE,
+        nonce: TEST_NONCE,
+        code_challenge: TEST_CODE_CHALLENGE,
+        code_challenge_method: 'S256',
+        connection: 'test-connection',
+        audience: 'test'
+      });
+    });
```

Comment:
Instead of verifying utils.createQueryParams was called with the correct params, we test the URL that is used to open the popup. This removes a level of mocking.

This pattern is used throughout the other tests below as wel.